src/SensitiveFloat.jl

```diff
@@ -0,0 +1,275 @@
+
+export multiply_sfs!, add_scaled_sfs!, combine_sfs!, add_sources_sf!
+
+@doc """
+A function value and its derivative with respect to its arguments.
+
+Attributes:
+  v:  The value
+  d:  The derivative with respect to each variable in
+      P-dimensional VariationalParams for each of S celestial objects
+      in a local_P x local_S matrix.
+  h:  The second derivative with respect to each variational parameter,
+      in the same format as d.  This is used for the full Hessian
+      with respect to all the sources.
+  hs: An array of per-source Hessians.  This will generally be reserved
+      for the Hessian of brightness values that depend only on one source.
+""" ->
+type SensitiveFloat{ParamType <: CelesteTypes.ParamSet, NumType <: Number}
+    # Actually a single value, but an Array to avoid memory allocation
+    v::Vector{NumType}
```

Comment:
The problem occurred in accumulate_source_brightness!.  I also recall that
before I refactored ElboDeriv, there was a construction where we defined

fsm = (fsm[1], fsm[2])

and then referred to

fsm[i].v

Is it possible that the construction of this tuple was causing some type
instability that led to this memory allocation?  I do think that not every
reference to a SF's value caused allocation.  This is interesting if
someone wants to try switching back and looking at memory allcoation, but I
do think there are bigger fish to fry before the SC deadline.





On Tue, Mar 8, 2016 at 12:46 PM, Kyle Barbary <notifications@github.com>
wrote:

> In src/SensitiveFloat.jl
> <https://github.com/jeff-regier/Celeste.jl/pull/137#discussion_r55424483>:
>
> > +A function value and its derivative with respect to its arguments.
> > +
> > +Attributes:
> > +  v:  The value
> > +  d:  The derivative with respect to each variable in
> > +      P-dimensional VariationalParams for each of S celestial objects
> > +      in a local_P x local_S matrix.
> > +  h:  The second derivative with respect to each variational parameter,
> > +      in the same format as d.  This is used for the full Hessian
> > +      with respect to all the sources.
> > +  hs: An array of per-source Hessians.  This will generally be reserved
> > +      for the Hessian of brightness values that depend only on one source.
> > +""" ->
> > +type SensitiveFloat{ParamType <: CelesteTypes.ParamSet, NumType <: Number}
> > +    # Actually a single value, but an Array to avoid memory allocation
> > +    v::Vector{NumType}
>
> Typically, if there's no type instability, modifying a bitstype field like
> Float64 should not allocate. Example:
>
> julia> type A
>        val::Float64
>        end
>
> julia> function update!(a)
>            for i=1:10_000_000
>                a.val = rand()
>            end
>        end
>
> julia> a = A(0.0);
>
> julia> @time update!(a)
>   0.048819 seconds (19.14 k allocations: 826.487 KB)
>
> julia> @time update!(a)
>   0.039082 seconds (4 allocations: 160 bytes)
>
> I don't know whether this is absolutely always true though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeff-regier/Celeste.jl/pull/137/files#r55424483>.
>